benchmarks/Benchmarks.hs

```diff
@@ -76,7 +76,7 @@ data Env = Env {
 
 setupEnv :: IO Env
 setupEnv = do
-    let n = 2^(12 :: Int)
+    let n = 4 * 2^(16 :: Int)
```

Comment:
It would be good to have a comment that explains how this number is chosen.

Reply:
Unfortunately, it seems that the increased number of elements causes some benchmarks to take ages now, for example:

```
benchmarking HashMap/isSubmapOf/Int ... took 533.9 s, total 5314525 iterations
benchmarked HashMap/isSubmapOf/Int
time                 172.1 ns   (122.6 ns .. 259.1 ns)
                     0.737 R²   (0.518 R² .. 0.915 R²)
mean                 202.4 ns   (192.3 ns .. 211.9 ns)
std dev              16.41 ns   (12.09 ns .. 19.93 ns)
variance introduced by outliers: 28% (moderately inflated)
```

What was the motivation for increasing this number, and is there another way to achieve that goal while keeping the total benchmarks runtime roughly within the original range?

Reply:
This benchmark takes much longer to run due to the number of iterations that gauge calculates to run. In this case gauge ran `5314525` iterations

> What was the motivation for increasing this number

For each level `i` we grow the tree by adding arrays of arrays, so at level 1 the HAMT can hold the base number of elements before it has to grow, and at level `i+1` it holds `base^{i+1}`. This means that to store `2^12 = 4096` elements, the 16-bit base version requires exactly `log_16(4096) = 3` levels and the 32-bit base version requires `log_32(4096) = 2.4` levels (so also 3). Thus with 4096 elements the 16-bit base version is in the best case scenario for most operations, especially if locality of reference is firing. If it is firing then we would have cache hits for the majority of accessing operations in the 16-bit version because each array will be fully loaded into the CPU caches, and since each array is full we will accidentally maximize CPU cache hits. However, for the 32-bit version this isn't the case, the 32bit version will still need 3 levels but with 4096 elements the HAMT will only be `4096 / (32^3) = 0.125 ~= 12%` full, and thus we'll have more CPU cache misses and a noisier benchmark.

So I decided to increase the number of elements to benchmark such that _both_ the 16-bit and 32-bit versions were not in a fully-packed scenario, and I wanted to increase the number of elements to exacerbate any difference between the two. I raised the element count to `2^16 = 65536` (which is still a best case for 16bit version) and then multiplied by 4 yielding `log_16(4*65536) = 4.5` levels for 16bit, and `log_32(4*65536) = 3.6` for 32-bit. With this element count, both the 16-bit version and the 32-bit version are at similar capacity:  16-bit: `(4 * 65536) / 16^5 ~= 25%`, 32-bit:  `(4 * 65536) / 32^4 ~= 25%`.

I'm sure other numbers would work, however it would probably require doing some math since getting the same capacity can be rather hard since one of the versions will either be full or significantly more full than the other. For example:
```
# 2^13
>>> (2**13) / (16**4)
0.125
>>> (2**13) / (32**3)
0.25

# 2^14
>>> (2**14) / (16**4)
0.25
>>> (2**14) / (32**3)
0.5

# 2^15
>>> (2**15) / (16**4)
0.5
>>> (2**15) / (32**3)
1.0

# even if we take some weird proportion one is still usually twice as full
>>> (2**15 / 3) / (16**4)
0.16666666666666666
>>> (2**15 / 3) / (32**3)
0.3333333333333333
```

In any case, the element count only matter for comparing the version in this PR. If the concern is CI or benchmark time then after this PR is merged (and the versions compared) we can set the element count back to 4096. In my opinion we shouldn't be benchmarking just the best case scenario anyhow.